classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,184 @@
+        $keepuserfield = $record->keepuserfield;
+        $keepuservalue = $record->keepuservalue;
+        $removeuserfield = $record->removeuserfield;
+        $removeuservalue = $record->removeuservalue;
+        // Verify user to remove.
+        $usertoremove = $DB->get_records(merge_request::TABLE_USERS, [$removeuserfield => $removeuservalue]);
+        if (is_null($record->removeuserid)) {
+            $this->removeuserid = $this->find_user_id_or_fail($record->removeuserfield, $record->removeuservalue);
+        }
+        // Verify user to keep.
+        $usertokeep = $DB->get_records(merge_request::TABLE_USERS, [$keepuserfield => $keepuservalue]);
+        if (is_null($record->keepuserid)) {
+            $this->keepuserid = $this->find_user_id_or_fail($record->keepuserfield, $record->keepuservalue);
+        }
+        return $returnvalue = 1;
+    }
+    /**
+     * Function for updating number of retries.
+     */
+    private function update_retries_in_table(int $idrecord,
+                                             int $retries) {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'retries' => $retries,
+                'timecompleted' >= time(),
+            ]
+        );
+    }
+    /**
+     * Function for updating removeuserid.
+     */
+    private function update_removeuserid_in_table(int $idrecord,
+                                                        int $removeuserid) {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'removeuserid' => $removeuserid
+            ]
+        );
+    }
+    /**
+     * Function for updating keepuserid.
+     */
+    private function update_keepuserid_in_table(int $idrecord,
+                                                int $keepuserid) {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'keepuserid' => $keepuserid
+            ]
+        );
+    }
+    /**
+     * Function to find user id or fail.
+     */
+    protected function find_user_id_or_fail(int $mergerequestid, string $userfield, string $uservalue): int {
+        global $DB;
+        $user = $DB->get_records(merge_request::TABLE_USERS, [$userfield => $uservalue]);
+        if (count($usertoremove) == 0) {
+            throw new moodle_exception(get_string('cannotfinduser', 'tool_mergeusers', (object)['userfield' => $userfield, 'uservalue' => $uservalue]));
+        } 
+        if (count($usertoremove) > 1) {
+            throw new moodle_exception(get_string('toomanyusers', 'tool_mergeusers', (object)['userfield' => $userfield, 'uservalue' => $uservalue]));
+        }
+        $this->update_removeuserid_in_table($mergerequestid , $user->id);
```

Comment:
`$user`, or `$users` as I proposed, is an array.

So, we should get the first and unique row of the array, like:

`$user = reset($users);`

so then you can really use `$user->id;`


Have you tried all this code while making the adjustments? This could not work.

Reply:
after all these updates I will try new code